### PR TITLE
Encode connection strings to escape semicolon etc

### PR DESCRIPTION
**Pull request by daybarr**
ODBC connection string values that contain a semicolon must be escaped, as the
semicolon is the separator used between KeyValuePairs. So build up the
connection string in a structured way, using a dict to store KeyValuePairs so
we can be sure to properly encode all the values used, even those that come
from Django settings where the raw, unescaped values will be given.

### CODE REVIEW

**Comment by daybarr** [comment]
We needed this change to allow us to use `;` or `{` in database passwords. Others might have similar issues with such characters in database name, user name, DSN etc.

There are two things about this change that I think are worth noting:
- Dropping the `os.path.isabs(driver)` check. This was added by #27 to support iODBC which presumably didn't like the driver path being wrapped in the `{}` escaping. Prior to #27, the escaping of the driver value was done all the time, regardless of whether or not it was necessary. With this change, the driver value is only escaped if it contains a semicolon, and most sensible file system paths do not contain semicolons. If a path to the driver did contain such a value, presumably iODBC would be unable to handle this anyway if it doesn't like the escaping.
- Restructuring the code to use a `dict` for `cstr_parts` instead of a list will result in the KeyValuePairs of the final connection string coming out in a different order. As order of KeyValuePairs is only relevant when there are repeated keys (see [[MS-ODBCSTR]](https://msdn.microsoft.com/en-us/library/ee208909%28v=sql.105%29.aspx)) this should not cause any problems, but there could be bad drivers out there that do rely on a specific ordering.

So in summary, these changes shouldn't cause any problems, but there is a small risk of oddness with badly behaved drivers and setups that I do not have available to test against. It would be possible to make a more minimal change to avoid any chance of these potential issues, but I think the resulting code is cleaner and more maintainable this way.

**Comment by michiya** [comment]
@daybarr Thank you for the contribution! I think it will work well on all ODBC drivers that the backend supports now, and this change will be included in the next release.

**Comment by daybarr** [comment]
@michiya No problem at all. Thank _you_ for all your efforts on this great project, please keep up the good work :)
